gr-uhd/lib/rfnoc_graph_impl.cpp

```diff
@@ -0,0 +1,193 @@
+{
+public:
+    rfnoc_graph_impl(const device_addr_t& dev_addr)
+        : _graph(::uhd::rfnoc::rfnoc_graph::make(dev_addr))
+    {
+        // nop
+    }
+
+    void connect(const std::string& src_block_id,
+                 const size_t src_block_port,
+                 const std::string& dst_block_id,
+                 const size_t dst_block_port,
+                 const bool skip_property_propagation)
+    {
+        if (_tx_streamers.count(src_block_id)) {
+            if (_rx_streamers.count(dst_block_id)) {
+                throw std::runtime_error("Cannot connect RFNoC streamers directly!");
+            }
+            _graph->connect(_tx_streamers.at(src_block_id),
+                            src_block_port,
+                            block_id_t(dst_block_id),
+                            dst_block_port);
+            return;
+        }
+        if (_rx_streamers.count(dst_block_id)) {
+            _graph->connect(src_block_id,
+                            src_block_port,
+                            _rx_streamers.at(dst_block_id),
+                            dst_block_port);
+            return;
+        }
+
+        _graph->connect(block_id_t(src_block_id),
+                        src_block_port,
+                        block_id_t(dst_block_id),
+                        dst_block_port);
+    }
+
+    void connect(const std::string& block1,
+                 const std::string& block2,
+                 bool skip_property_propagation)
+    {
+        connect(block1, 0, block2, 0, skip_property_propagation);
+    }
+
+    void connect(rfnoc_block::sptr src_block,
+                 const size_t src_block_port,
+                 rfnoc_block::sptr dst_block,
+                 const size_t dst_block_port,
+                 const bool skip_property_propagation)
+    {
+        connect(src_block->get_unique_id(),
+                src_block_port,
+                dst_block->get_unique_id(),
+                dst_block_port,
+                skip_property_propagation);
+    }
+
+
+    ::uhd::rx_streamer::sptr create_rx_streamer(const size_t num_ports,
+                                                const stream_args_t& args)
+    {
+        auto streamer = _graph->create_rx_streamer(num_ports, args);
+        const std::string streamer_id =
+            std::dynamic_pointer_cast<::uhd::rfnoc::node_t>(streamer)->get_unique_id();
+        _rx_streamers.insert({ streamer_id, streamer });
+        return streamer;
+    }
+
+    ::uhd::tx_streamer::sptr create_tx_streamer(const size_t num_ports,
+                                                const stream_args_t& args)
+    {
+        auto streamer = _graph->create_tx_streamer(num_ports, args);
+        const std::string streamer_id =
+            std::dynamic_pointer_cast<::uhd::rfnoc::node_t>(streamer)->get_unique_id();
+        _tx_streamers.insert({ streamer_id, streamer });
+        return streamer;
+    }
+
+    void commit()
+    {
+        if (!_commit_called.exchange(true)) {
```

Comment:
Interesting use of `exchange()`--do you expect `commit()` to be called from multiple threads and are trying to protect against multiple commits?

Reply:
If so, is that the only call that can get called from multiple threads? Do the rest of the functions that mutate state also need to be concerned?

Reply:
commit() will be called from all the streamers (at least, we might add more commits in the future). The block controllers are fine, they have mutexes either in UHD, or they are uniquely accessing a block controller. rfnoc_graph is the only one that gets accessed from many threads, and I think I have all the mutexes in place.

Reply:
Okay, works for me!